### PR TITLE
[Fix #5289] Fix conflict between SpaceInsideBrackets cops

### LINKED ISSUE #5289: Infinite loop with Layout/SpaceInsideArrayLiteralBrackets and Layout/SpaceInsideReferenceBrackets
## Expected behavior

Here's the configuration:

```
Layout/SpaceInsideArrayLiteralBrackets:
  EnforcedStyle: space
```

And here's the code:

```
a = []
a[[ 1, 2 ]]
```

I expect this to pass. SpaceInsideArrayLiteralBrackets should be happy since there is a space inside the brackets. SpaceInsideReferenceBrackets should be happy since there are no spaces in between the `[[` and `]]`. It should complain if I were doing `a[ [ 1, 2 ] ]`, but I'm not.

## Actual behavior

When I use auto-correct, the two cops fight each other and create an infinite loop.

## Steps to reproduce the problem

Create the files above as `style.yml` and `a.rb`, then run:

```
rubocop -c style.yml a.rb  --auto-correct
```

## RuboCop version

0.52.0 (using Parser 2.4.0.2, running on ruby 2.2.2 x86_64-linux)

### PR DESCRIPTION
**Pull request by garettarrowood**
Resolves infinite loop issue described in #5289 and adds integration tests for `Layout/SpaceInsideReferenceBrackets` and `Layout/SpaceInsideArrayLiteralBrackets`.

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* [x] Run `rake default` or `rake parallel`. It executes all tests and RuboCop for itself, and generates the documentation.

[1]: http://chris.beams.io/posts/git-commit/

### CODE REVIEW

**Comment by garettarrowood** [comment]
CI currently red without the fix from #5299 .  Before a rebase for a CHANGELOG merge conflict, this ran green.

File: lib/rubocop/cop/layout/space_inside_reference_brackets.rb [hunk 1]
```diff
@@ -65,16 +65,24 @@ def autocorrect(node)
...
         end
 
-        def right_ref_bracket(node, token)
-          i = tokens(node).index(token)
-          tokens(node).slice(i..-1).find(&:right_bracket?)
+        def closing_bracket(node, opening_bracket)
+          i = tokens(node).index(opening_bracket)
+          inner_left_brackets_needing_closure = 0
+
+          tokens(node).slice(i..-1).each do |token|
```
Comment (Drenmi) [inline lib/rubocop/cop/layout/space_inside_reference_brackets.rb]: Wonder if there's any reason for using `#slice` over `#[]` here? 🤔

Comment (garettarrowood) [reply lib/rubocop/cop/layout/space_inside_reference_brackets.rb]: Good call!

Comment (garettarrowood) [reply lib/rubocop/cop/layout/space_inside_reference_brackets.rb]: Might this be a good cop candidate? We'd have to make sure the receiver is an array.

**Pull request merged**